### PR TITLE
Support JSON streaming with Multi

### LINKED ISSUE #14183: Clarify use of Multi as return type in RESTEasy Reactive
**Description**
Currently RESTEasy Reactive always tries to "stream" a result when `Multi` is used a return type of Resource method.

This can be surprising in cases like the following:

```java
@Path("fruits")
@ApplicationScoped
@Produces("application/json")
@Consumes("application/json")
public class FruitMutinyResource {
    private static final Logger LOGGER = Logger.getLogger(FruitMutinyResource.class.getName());

    @Inject
    Mutiny.Session mutinySession;

    @GET
    public Multi<Fruit> get() {
        return mutinySession
                .createNamedQuery( "Fruits.findAll", Fruit.class ).getResults();
    }

}
```

where users expect to get a JSON array as a result, but instead end up with something like

```json
{"id":2,"name":"Apple"}{"id":3,"name":"Banana"}{"id":1,"name":"Pineapple"}
```

We have the following options:

* Keep things as they are (as in `Multi` always "streams") and expect users to collect the results to `Uni` (using `.collectItems().asList()` on the `Multi`)
* Have RESTEasy Reactive automatically collect the result to `Uni` when the media type of resource method does not support streaming (for example `application/json` does not stream). This may not be bullet-proof however as there is no spec mandate about which media types can and can't stream.

**Historical note**

The discussion that lead to this is [here](https://github.com/quarkusio/quarkus/issues/14183)

### PR DESCRIPTION
**Pull request by FroMage**
It's all super hackish, but better than the current behaviour of producing invalid json

Relates to #14183

### CODE REVIEW

**Review by geoand — APPROVED** [review]
This is pretty hacky, but it's  better than what we have now.

Let's keep the issue open however so we can track the need for a proper solution

File: independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/MultiResponseHandler.java [hunk 1]
```diff
@@ -42,30 +43,65 @@ public Object apply(Object v, Throwable t) {
...
+                                // need to cancel because the exception didn't come from the Multi
+                                try {
+                                    subscription.cancel();
+                                } catch (Throwable t2) {
+                                    t2.printStackTrace();
+                                }
+                                handleException(requestContext, t);
+                            } else {
+                                // next item will need this prefix if json
+                                nextJsonPrefix = ",";
```
Comment (cescoffier) [inline independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/MultiResponseHandler.java:75]: Aren't you adding a `,` even for the last item?

Comment (FroMage) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/MultiResponseHandler.java:75]: No, this is only used if we write an item, not at the end of the stream.

**Comment by FroMage** [comment]
> This is pretty hacky, but it's better than what we have now.

Super hacky. But the alternative is to disallow JSON streaming, or to special-case it and collect it before sending.

I _suppose_ the fact that we can actually stream it this way is better because it opens up the possibility of using that streaming from the client-side.

**Comment by cescoffier** [comment]
There is a case not covered by tests that need to be documented.
Imagine something like:

```java
@GET
@Path("/multiX")
public Multi<Person> getMultiX() {
    Person person = new Person();
    person.setFirst("Bob");
    person.setLast("Builder");
    return Multi.createFrom().items(person)
      .onCompletion().failWith(() -> new Exception("boom"));
}
```

Basically, a failure happening during the streaming.
You cannot set the status to 500, as the headers have already been written (BTW, if the failure happens before having sent any item, it should be the case).

What should be the outcome in this case? Should we encode the failure in JSON followed by `]` and let the client handle the failure? Should we just end the response abruptly (no ending `]`)?

**Comment by FroMage** [comment]
Yeah, this is just not specified. Neither SSE nor streaming support that sort of thing, so we'd have to decide what we want to do. But that's another issue for you to file please :)

**Comment by cescoffier** [comment]
@FroMage https://github.com/quarkusio/quarkus/issues/14222

File: independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/MultiResponseHandler.java (hunk 1 above)
Comment (cescoffier) [reply independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/MultiResponseHandler.java:75]: oh sorry! Got it!

**Comment by gsmet** [comment]
Do we want this in 1.11? If so, please add the `triage/backport?` label.

**Comment by geoand** [comment]
I think we should have it. @FroMage WDYT?

**Comment by FroMage** [comment]
Yup, done

**Pull request merged**